field_utils/FieldPointerUtilities.F90

```diff
@@ -796,159 +837,185 @@ subroutine copy_r8_r8(cptr_x, cptr_y, n)
+         isPresent = ESMF_InfoIsPresent(infoh,"missing_value",_RC)
+         _ASSERT(isPresent,"missing undef value")
+         call ESMF_InfoGet(infoh,value=undef_values(i),key="missing_value",_RC)
+      enddo
+      _RETURN(_SUCCESS)
+   end subroutine GetFieldsUndef_r4
+
+   subroutine GetFieldsUndef_r8(fields,undef_values,rc)
+      type(ESMF_Field), intent(inout) :: fields(:)
+      real(kind=ESMF_KIND_R8), allocatable,intent(inout) :: undef_values(:)
+      integer, optional, intent(out) :: rc
+
+      integer :: status, i
+      logical :: isPresent
+      type(ESMF_Info) :: infoh
+
+      allocate(undef_values(size(fields)))
+      do i =1,size(fields)
+         call ESMF_InfoGetFromHost(fields(i),infoh,_RC)
+         isPresent = ESMF_InfoIsPresent(infoh,"missing_value",_RC)
+         _ASSERT(isPresent,"missing undef value")
+         call ESMF_InfoGet(infoh,value=undef_values(i),key="missing_value",_RC)
+      enddo
+      _RETURN(_SUCCESS)
+   end subroutine GetFieldsUndef_r8
+
+   subroutine Destroy(Field,RC)
+      type(ESMF_Field),          intent(INOUT) :: Field
+      integer, optional,         intent(OUT  ) :: RC
+
+      integer                               :: STATUS
+
+      real(kind=ESMF_KIND_R4), pointer      :: VR4_1D(:), VR4_2D(:,:), VR4_3D(:,:,:), VR4_4D(:,:,:,:)
+      real(kind=ESMF_KIND_R8), pointer      :: VR8_1D(:), VR8_2D(:,:), VR8_3D(:,:,:), VR8_4D(:,:,:,:)
+      integer                      :: rank
+      type(ESMF_TypeKind_Flag)     :: tk
+      logical :: esmf_allocated
+
+      call ESMF_FieldGet(Field,typekind=tk,dimCount=rank,isESMFAllocated=esmf_allocated,_RC)
+      if (.not. esmf_allocated) then
+         if (tk == ESMF_TYPEKIND_R4 .and. rank == 1) then
+            call ESMF_FieldGet(Field,0,VR4_1d,_RC)
+            deallocate(VR4_1d,_STAT)
+         else if (tk == ESMF_TYPEKIND_R8 .and. rank == 1) then
+            call ESMF_FieldGet(Field,0,VR8_1d,_RC)
+            deallocate(VR8_1d,_STAT)
+         else if (tk == ESMF_TYPEKIND_R4 .and. rank == 2) then
+            call ESMF_FieldGet(Field,0,VR4_2d,_RC)
+            deallocate(VR4_2d,_STAT)
+         else if (tk == ESMF_TYPEKIND_R8 .and. rank == 2) then
+            call ESMF_FieldGet(Field,0,VR8_2d,_RC)
+            deallocate(VR8_2d,_STAT)
+         else if (tk == ESMF_TYPEKIND_R4 .and. rank == 3) then
+            call ESMF_FieldGet(Field,0,VR4_3D,_RC)
+            deallocate(VR4_3d,_STAT)
+         else if (tk == ESMF_TYPEKIND_R8 .and. rank == 3) then
+            call ESMF_FieldGet(Field,0,VR8_3D,_RC)
+            deallocate(VR8_3d,_STAT)
+         else if (tk == ESMF_TYPEKIND_R4 .and. rank == 4) then
+            call ESMF_FieldGet(Field,0,VR4_4D,_RC)
+            deallocate(VR4_3d,_STAT)
+         else if (tk == ESMF_TYPEKIND_R8 .and. rank == 4) then
+            call ESMF_FieldGet(Field,0,VR8_4D,_RC)
+            deallocate(VR8_3d,_STAT)
+         else
+            _FAIL( 'unsupported typekind+rank')
+         end if
+      end if
```

Comment:
I know this is not part of this PR, but ... MAPL cannot delete this here for the same reason that ESMF cannot in their Destroy call.   When the user has allocated the memory, there is no way to know if something else is using the memory.    


Do not change now, but please discuss with me.  Let's be sure we have a use case for this.

Reply:
I understand. I don't know why those are deleted. The only reason that it shows up as a change is the indentation. I ran a vim command to fix the indentation because it was incorrect and inconsistent. It should follow the 3 space convention. I am not familiar with that code.